nixos/*: split docs build

pennae opened the pull request:
###### Motivation for this change

almost all options documentation is constant, only very few options reference `pkgs` in attributes that end up in documentation. this fact can be used to build *most* of options doc in a derivation that can then be cached, and merging the rest of the docs (which can evaluated more cheaply now) separately.

doing so reduces nixos-rebuild time quite significantly (by 20% or more), reducing build time on our system from 9.4 seconds to 7.4 seconds. disabling docs entirely reduces build time to 7.1 seconds, making the docs build as close to free as we can currently get.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

pennae committed:
nixos/documentation: split options doc build

most modules can be evaluated for their documentation in a very
restricted environment that doesn't include all of nixpkgs. this
evaluation can then be cached and reused for subsequent builds, merging
only documentation that has changed into the cached set. since nixos
ships with a large number of modules of which only a few are used in any
given config this can save evaluation a huge percentage of nixos
options available in any given config.

in tests of this caching, despite having to copy most of nixos/, saves
about 80% of the time needed to build the system manual, or about two
second on the machine used for testing. build time for a full system
config shrank from 9.4s to 7.4s, while turning documentation off
entirely shortened the build to 7.1s.

pennae committed:
nixos/kubernetes: add missing defaultText to expression default

pennae committed:
nixos/kubernetes: make lib option internal and readonly

this set almost certainly shouldn't be touched by users, nor listed in
the manual. make it internal and use it only through the option path to
make clear that this should not be modified.

pennae committed:
nixos/couchdb: add missing defaultText

pennae committed:
nixos/xrdp: add missing defaultText

pennae committed:
nixos/aesmd: add missing defaultText

pennae commented:
most of the modules that have gained an exemption from sandboxed doc builds either interpolate the version of the package they use into doc attributes, or they set `relatedPackages`. the former isn't easily fixed, the latter could be fixed by separating `relatedPackages` info from the modules themselves or removing the feature altogether. separating probably doesn't gain us much before there are a lot of users of this feature, but since it's been around for quite a while and hasn't seen much uptake that doesn't seem likely. (and if many users *did* appear we'd probably lose all the build time improvements the split provices, so maybe it'd be better to remove `relatedPackages` info as it exists now?)

roberth commented on nixos/modules/misc/documentation.nix:
```diff
@@ -36,6 +60,43 @@ let
+            --arg modules "[ $modules ]" \
+            --argstr stateVersion "${options.system.stateVersion.default}" \
+            --argstr release "${config.system.nixos.release}" \
+            ${./eval-cacheable-options.nix} > $out \
```
This adds this file to the store independently, but it is already in `${toString pkgs.path}` so you could use that instead.

pennae replied:
nope, that's *much* more expensive than the many small adds this does. :( would be nice if there was a way to tell nix to reuse an existing store path, much like it does for derivations.

roberth replied:
If you reuse `$nixosPath` you don't need to add anything extra to the store, right?

pennae replied:
ah, think i misunderstood. yeah, we can totally use `$nixosPath/modules/misc/eval-cacheable-options.nix` instead!

cole-h commented on nixos/modules/misc/documentation.nix:
```diff
@@ -1,30 +1,54 @@
-{ config, lib, pkgs, extendModules, noUserModules, ... }:
+{ config, lib, options, pkgs, baseModules, extraModules, modules, modulesPath, utils, ... }:
```
Extremely minor nit, but since `config` and `options` are created in sequence in most modules (e.g. `config = {};` and then `options = {}`), and since I'd put `utils` and `modules` as the same "priority" as `config`, `options`, etc (mostly just because they're all lowercase, not snake case)...:

```suggestion
{ config, options, lib, pkgs, utils, modules, baseModules, extraModules, modulesPath, ... }:
```

roberth replied:
I either sort alphabetically or don't care.

roberth commented on nixos/modules/misc/documentation.nix:
```diff
@@ -1,30 +1,54 @@
-          specialArgs.pkgs = scrubDerivations "pkgs" pkgs;
+        scrubbedEval = evalModules {
+          modules = [ {
+            nixpkgs.localSystem = config.nixpkgs.localSystem;
```
I don't think you should need this for docs.

Duplicating the `evalModules` logic is not great for maintainability. Would it be feasible to use `extendModules` when module filtering is supported by the module system itself?

pennae replied:
you're right, that isn't needed here. wonder where that even came from now, probably from attempts to put `nixpkgs.nix` into the sandbox (which failed anyway) :thinking:

roberth replied:
You could do essentially `extendModules { modules = [ { config._module.args.check = lib.mkForce false; } ]; }` for `check`.
The only challenge is removing some modules.

pennae replied:
"some" is a bit of an understatement, there's over 1400 of them at time of writing :smile: filtering loaded modules like that would be possible if that facility existied, but evaluating a few user modules from scratch is probably faster. :thinking:

cole-h reviewed (commented):
Finally making good on my 2-week-old promise...

pennae commented on nixos/modules/misc/documentation.nix:
```diff
@@ -36,6 +60,43 @@ let
+            );
+      in
+        pkgs.runCommand "lazy-options.json" {
+          libPath = filter "${toString pkgs.path}/lib";
+          pkgsLibPath = filter "${toString pkgs.path}/pkgs/pkgs-lib";
+          nixosPath = filter "${toString pkgs.path}/nixos";
```
copying the full `pkgs.path` is really expensive, to the point where it's not worth the effort of splitting the docs build at all. if there was a "use this store path exactly without copying" we'd love to use that instead of copying peacemeal, but that wouldn't help use cases that pin nixpkgs to a non-store path either :(

roberth replied:
My understanding of this code was wrong at first.
I think we can optimize the case of an already in-store `pkgs.path`, such as with flakes, niv or really any nix-based pinning method. In this case, we don't have to copy anything, as the path is already in the store, ready to be used in a build. e.g.

```nix
if lib.isStorePath pkgs.path
  then "${builtins.storePath pkgs.path}/lib"
  else filter "${toString pkgs.path}/lib"
```

`builtins.storePath` is supported under both pure and restricted evaluation. It basically adds the store path to the context rather than adding a _new_ directory to the store.

pennae replied:
nice! that's exactly what we were looking for and missed. the docs do however say that it's *not* available in pure evaluations, but it seems to work fine for flakes builds regardless. thanks for the pointer! :+1:

roberth commented:
> could be fixed by separating `relatedPackages` info from the modules themselves or removing the feature altogether.

I find it surprising that a bunch of mere strings are causing such an issue. Perhaps the feature could be less strict, though I wouldn't have noticed in practice if you removed it.

pennae commented on nixos/lib/make-options-doc/mergeJSON.py:
```diff
@@ -0,0 +1,71 @@
+            cur[ok] = ov
+
+# check that every option has a description
+# TODO: nixos-rebuild with flakes may hide the warning, maybe turn on -L by default for those?
```
we could redirect them to another file in the `options.json` out path, but that then has to be propagated at least five steps deep to make it to activation. would prefer to turn the warnings into errors unless the user explicitly turns them off to be honest.

roberth replied:
Propagating the warnings to evaluation would be IFD. Let's keep it simple and turn it into an error.

Also doing it in activation is far from ideal, because now these originally static checks have to run at deployment time, or in case of dry activation, not deployment time but far too late into the process and on the wrong machine (considering remote activation such as NixOps, [runNixOS](https://docs.hercules-ci.com/hercules-ci-effects/reference/nix-functions/runnixos/), etc)

pennae replied:
the warnings are now errors by default, but with an option to turn them into mere warnings (so as to not break builds that worked previously *too* much)

pennae commented:
> I find it surprising that a bunch of mere strings are causing such an issue. Perhaps the feature could be less strict, though I wouldn't have noticed in practice if you removed it.

it's not the strings that cause the issue, but how they are handled. the docs mechanism uses those strings as attrnames into `pkgs` and pulls the package description into the docs.

roberth commented:
> > I find it surprising that a bunch of mere strings are causing such an issue. Perhaps the feature could be less strict, though I wouldn't have noticed in practice if you removed it.
> 
> it's not the strings that cause the issue, but how they are handled. the docs mechanism uses those strings as attrnames into `pkgs` and pulls the package description into the docs.

Perhaps these could be replaced by links to `search.nixos.org` or something. Using the `show=` query param, example: https://search.nixos.org/packages?show=hercules-ci-agent&sort=relevance&query=hercules-ci-agent

pennae commented:
turning them into links sounds like a good idea. the manpage build currently renders links in bold red, if we set the link text to something like `pkgs.<attribute name>` and dropped the package metadata we'd get a nice html manual and the manpage would be sufficient to look up information. does that sound sensible?

roberth commented:
Yeah, go for it :)

pennae committed:
nixos/make-options-doc: add type annotations to mergeJSON.py

pennae committed:
nixos/make-options-doc: turn relatedPackages into links

link to search.nixos.org instead of pulling package metadata out of pkgs. this
lets us cache docs of a few more modules and provides easier access to package
info from the HTML manual, but makes the manpage slightly less useful since
package description are no longer rendered.

roberth commented on nixos/modules/misc/documentation.nix:
```diff
@@ -1,30 +1,53 @@
+    in
+      {
+        lazy = p.right;
+        eager = p.wrong ++ optionals cfg.nixos.includeAllModules (extraModules ++ modules);
```
For `includeAllModules`, you'd only have to add the user-provided modules, right? (in addition to `p.wrong`)

pennae replied:
not sure. `extraModules` was included in the docs build when we started, so we just kept it.

as far as we understand `extraModules` (via NIXOS_EXTRA_MODULE_PATH) are always loaded but not included by default for performance reasons. it would be nice to include them in the cached part of the build, but like user modules they can reference arbitrary paths so that didn't work out :(

pennae committed:
nixos/make-options-doc: treat missing descriptions as errors by default

this partially solves the problem of "missing description" warnings of the
options doc build being lost by nix build, at the cost of failing builds that
previously ran. an option to disable this behaviour is provided.

pennae commented on nixos/lib/make-options-doc/mergeJSON.py:
```diff
@@ -0,0 +1,71 @@
+Option = collections.namedtuple('Option', ['name', 'value'])
+
+# pivot a dict of options keyed by their display name to a dict keyed by their path
+def pivot(options):
```
added some type annotations and mypy seems to like them, but we're not python exports. are these sufficient?

pennae committed:
nixos/documentation: avoid copying nixpkgs subpaths

the docs build should work well even when called from a git checkout of
nixpkgs, but should avoid as much work as possible in all cases.
if pkgs.path is already a store path we can avoid copying parts of it
into the docs build sandbox by wrapping pkgs.path in builtins.storePath

cole-h commented on nixos/lib/make-options-doc/mergeJSON.py:
```diff
@@ -0,0 +1,86 @@
+        result[opt.name] = opt.value
+    return result
+
+warningsAreErrors = sys.argv[1] == "-Werror"
```
As funny as this is, I think I'd prefer to see `--treat-warnings-as-errors` or `--warnings-are-errors` or something similarly verbose. I don't think this will be used much (if at all) outside of the related tooling here, so I think it's a good idea to be verbose enough that anyone could understand what it's doing without having experience with compiler flags.

pennae replied:
yeah, that's fair. will change it post haste, once the rebase to master finishes :)

cole-h reviewed (commented):
One comment, else the rest LGTM.

pennae committed:
nixos/sourcehut: add missing defaultText, escape antiquotations

pennae commented:
renamed the warnings-as-errors flag of mergeJSON.py, rebased onto current master: sourcehut was missing a few defaultTexts. no other changes.

pennae commented:
we'd say this is ready for merge. @roberth what do you think, is there anything that still needs doing?

roberth commented on nixos/lib/make-options-doc/mergeJSON.py:
```diff
@@ -0,0 +1,86 @@
+if hasWarnings and warningsAreErrors:
+    print(
+        "\x1b[1;31m" +
+        "Treating warnings as errors. Set documentation.nixos.optionsDoc.warningsAreErrors " +
```
```suggestion
        "Treating warnings as errors. Set documentation.nixos.options.warningsAreErrors " +
```

`Doc` is redundant. Please rename both new options.

pennae replied:
renamed both. :)

roberth commented on nixos/doc/manual/development/meta-attributes.section.md:
````diff
@@ -38,3 +39,12 @@ file.
     $ nix-build nixos/release.nix -A manual.x86_64-linux
     ```
+
+-  `buildDocsInSandbox` determines whether the option documentation
````
```suggestion
-  `buildDocsInSandbox` indicates whether the option documentation
```

roberth commented on nixos/doc/manual/development/meta-attributes.section.md:
```diff
@@ -38,3 +39,12 @@ file.
+   in its documentation (e.g. because it loads information from a linked
+   package to build an option type) or if its documentation depends on
+   other modules that also aren't sandboxed (e.g. by using types defined
+   in the other module).
```
Please explain why the sandbox is used and why it imposes restrictions on option attributes.

roberth commented on nixos/modules/misc/documentation.nix:
```diff
@@ -36,6 +59,48 @@ let
+            $nixosPath/lib/eval-cacheable-options.nix > $out \
+            || {
+              echo -en "\e[1;31m"
+              echo "cacheable portion of option doc build failed."
+              echo "usually this means that an option attribute that ends up in documentation (eg"
+              echo "\`default\` or \`description\`) depends on the restricted module arguments "
+              echo "\`config\` or \`pkgs\`."
```
This is a crucial error message that should be actionable.
Please add how to resolve such an error or link to a section in the documentation that explains it.

roberth reviewed (commented):
Technically looks good to me, but needs to be a bit more user friendly.

It will likely break a few custom out-of-tree modules, so let's make sure that these people will understand the change and why it is good for them.

pennae commented:
out-of-tree modules, like other user modules, are not built inside the sandbox; that is reserved for nixpkgs modules. non-nixpkgs modules also get to see all options that are currently loaded in the system, so no docs build breakages *should* (knock wood) happen.

roberth commented:
> out-of-tree modules, like other user modules, are not built inside the sandbox; that is reserved for nixpkgs modules. non-nixpkgs modules also get to see all options that are currently loaded in the system, so no docs build breakages _should_ (knock wood) happen.

Right, that's a good measure that I honestly forgot about, though modules in forks are susceptible.

pennae commented:
expanded documentation and added some suggestions to the error message. hope they are clear enough to be useful.

pennae commented:
anything else we could do to make this any more awesome? :) (would merge it now if we could, two approvals seems good.)

ajs124 commented:
This (bisect says fc614c37c653637e5475a0b0a987489b4d1f351d) seems to have broken the installer tests, therefore blocking the nixos channels.

The test fails with:
```
machine # [    8.959436] dhcpcd[827]: eth0: adding default route via fe80::2
machine # error: path '/nix/store/j02qf8rza5g930jazw4k71g0nigc063f-nixos-22.05pre-git' is not valid
machine # (use '--show-trace' to show detailed location information)
machine: output: 
Test "Perform the installation" failed with error: "command `nixos-install < /dev/null >&2` failed (exit code 1)"
```

From what I remember the installer tests use `closureInfo` to populate the store in the vm. So some part of that probably fails now.

vcunat commented:
So (partial?) revert for now to unblock channels?

pennae commented:
https://github.com/NixOS/nixpkgs/pull/153620 has a proposed fix. if that can't be merged quickly let's revert https://github.com/NixOS/nixpkgs/commit/fc614c37c653637e5475a0b0a987489b4d1f351d

vcunat commented:
Unfortunately that still doesn't completely fix the regressions caused by this PR on `nixosTests.installer.simple`, so channel blocking continues.  Some dependencies are now unreachable from within the installer tests, giving lines like:
```
machine # error: unable to download 'https://salsa.debian.org/debian/keyutils/raw/4cecffcb8e2a2aa4ef41777ed40e4e4bcfb2e5bf/debian/patches/Make-build-reproducible.patch
```

vcunat commented:
Usually these were addressed by extending `system.extraDependencies` in `nixos/tests/installer.nix` with the right packages (not _their_ build_ dependencies which happen to show up).   But I haven't looked into what exactly is newly required here and if it might be an undesirable addition instead.

pennae commented:
d'oh. didn't let the installer test run through, so we didn't see those. will open a PR to disable the docs split for now until the installer issues are fully sorted.

vcunat commented:
What I see is certainly just a test-only problem, so perhaps disable in the installer tests only?

vcunat commented:
What I see failing certainly are dependencies to generate docs (docbook, manual, manpages).

pennae commented:
we had tried disabling docs in the installer yesterday and had no success. could've been because tiredness, but at the moment i'd rather disable the split build to unblock the channel and have more time to analyse things.

zowoq commented:
7e28421e1704c95c056f2b2e7fc27a7569182e0f broke `kubernetes` as well.

zowoq commented:
And I can't even revert it because it breaks the manual.

pennae commented:
@zowoq does it unbreak when reverting the kubernetes commits and setting `meta.buildDocsInSandbox = false` for all kubernetes modules?

pennae commented:
seems so, the test does complete and the manual builds. #153951

bb010g commented:
This PR broke my NixOS configuration (with modules that customize existing modules) when switching from 21.11 to 22.05, and I had to bisect (`git bisect start release-22.05 release-21.11 && git bisect run nix-instantiate nixos -A system`) to find this PR.

Here's a sample evaluation failure:

```
error: The option `security.pam.services.<name>.text' does not exist. Definition values:
       - In `/etc/nixos/modules/modules/pam.nix':
           {
             _type = "override";
             content = ''
               session optional ${pkgs.pam}/lib/security/pam_umask.so 0022
             '';
           ...

       … while evaluating the attribute 'options'
```

I'm currently trying to figure out how to make this build again without turning `nixos.splitOptionDocBuild` off. It looks like `specialArgs = { options = allOpts; …; }` isn't working as intended.

pennae commented:
can you share a larger example that causes the errors you're seeing? could be that the doc split optimization simply isn't applicable in your case and turning it off could be the only option.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/rebuilding-system-using-niv-fails-in-nixos-22-05/19667/7

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/a-look-at-nixos-nixpkgs-evaluation-times-over-the-years/65114/20
